### ISSUE TITLE
Keyboards on UHCI controllers become unresponsive if typing too fast

### ISSUE THREAD

**Issue by debrouxl** [memtest86plus/memtest86plus#51]
Followup to #33 : on two ~20 year old MSI MBs of mine with UHCI controllers, USB keyboards now work... as long as one does not enter keypresses too quickly :)
Beyond some rate, above one keypress per refresh step of the memory test, keypresses stop being taken into account on the given keyboard. There's no recovery of the functionality within a reasonable amount of time.

Like #33 , this isn't high priority, since the vast majority of the MBs of that era do have a PS/2 port - that's what I use to reboot the computer in the end. However, I'm reporting this because two MBs which have little in common besides the manufacturer are affected.

* MSI KT4AV (MS-6712) with Athlon XP 3000+, 333 MHz bus, 2 x 512 MB DDR-333:
```
00:10.0 USB controller: 1106:3038 VIA Technologies, Inc. VT82xx/62xx/VX700/8x0/900 UHCI USB 1.1 Controller (rev 80)
00:10.1 USB controller: 1106:3038 VIA Technologies, Inc. VT82xx/62xx/VX700/8x0/900 UHCI USB 1.1 Controller (rev 80)
00:10.2 USB controller: 1106:3038 VIA Technologies, Inc. VT82xx/62xx/VX700/8x0/900 UHCI USB 1.1 Controller (rev 80)
00:10.3 USB controller: 1106:3104 VIA Technologies, Inc. USB 2.0 EHCI-Compliant Host-Controller (rev 82)
```

* MSI 845E Max2 (MS-6398E) with Northwood P4 2.6 GHz / 400 MHz FSB but effectively 266 MHz bus, 512 MB DDR-266 (memtest86+ says DDR-285, FWIW) & 512 MB DDR-333:
```
00:1d.0 USB controller: 8086:24c2 Intel Corporation 82801DB/DBL/DBM (ICH4/ICH4-L/ICH4-M) USB UHCI Controller #1 (rev 01)
00:1d.1 USB controller: 8086:24c4 Intel Corporation 82801DB/DBL/DBM (ICH4/ICH4-L/ICH4-M) USB UHCI Controller #2 (rev 01)
00:1d.2 USB controller: 8086:24c7 Intel Corporation 82801DB/DBL/DBM (ICH4/ICH4-L/ICH4-M) USB UHCI Controller #3 (rev 01)
00:1d.7 USB controller: 8086:24cd Intel Corporation 82801DB/DBM (ICH4/ICH4-M) USB2 EHCI Controller (rev 01)
```

**Comment by martinwhitaker** [comment]
I used this patch to try to reproduce this issue:
````
--- a/system/usbhcd.c
+++ b/system/usbhcd.c
@@ -742,7 +742,12 @@ void find_usb_keyboards(bool pause_if_none)
 
     if (usb_init_options & USB_DEBUG) {
         print_usb_info("Press any key to continue...");
-        while (get_key() == 0) {}
+//      while (get_key() == 0) {}
+        char ch;
+        while ((ch = get_key()) != '\n') {
+            if (ch != 0) print_char(print_row, print_col++, ch);
+            sleep(1);
+        }
     } else if (pause_if_none && num_usb_controllers == 0) {
         for (int i = PAUSE_IF_NONE_TIME; i > 0; i--) {
             print_usb_info("No USB keyboards found. Continuing in %i second%c ", i, i == 1 ? ' ' : 's');
````
but without success when using QEMU to emulate a UHCI controller with pass-through from a real K120 keyboard. Typing as rapidly as I could, I got one character echoed per second, which is what I expect as the USB keyboard does not buffer key presses.

If the above patch allows you to reproduce the issue, try adding this to see if it provides any clues:
````
--- a/system/uhci.c
+++ b/system/uhci.c
@@ -379,8 +379,10 @@ static uint8_t get_keycode(const usb_hcd_t *hcd)
     uint16_t io_base = ws->io_base;
 
     uint32_t status = inw(UHCI_USBSTS) & (UHCI_USBSTS_INT | UHCI_USBSTS_ERR);
+#if 0
     if (status != 0) {
         outw(UHCI_USBSTS_INT | UHCI_USBSTS_ERR, UHCI_USBSTS);
+#endif
 
         for (int kbd_idx = 0; kbd_idx < ws->num_keyboards; kbd_idx++) {
             uhci_qh_t *kbd_qh = &ws->qh[1 + kbd_idx];
@@ -388,6 +390,7 @@ static uint8_t get_keycode(const usb_hcd_t *hcd)
 
             status = kbd_td->control_status;
             if (status & UHCI_TD_ACTIVE) continue;
+            print_usb_info("kbd %i status %08x", kbd_idx, (uintptr_t)status);
 
             hid_kbd_rpt_t *kbd_rpt = &ws->kbd_rpt[kbd_idx];
 
@@ -408,7 +411,9 @@ static uint8_t get_keycode(const usb_hcd_t *hcd)
             write32(&kbd_td->control_status, kbd_td->driver_data[0]);
             write32(&kbd_qh->qe_link_ptr, (uintptr_t)kbd_td | UHCI_LP_TYPE_TD);
         }
+#if 0
     }
+#endif
 
     int kc_index_o = ws->kc_index_o;
     if (kc_index_o != ws->kc_index_i) {
````

**Comment by debrouxl** [comment]
I can reproduce the issue with the first patch, more easily on the MS-6712, where the flow of characters stops more quickly and more abruptly, than on the MS-6398E, where the flow of characters can occasionally resume for a short while and stop again. The stop seems to occur after spamming the keyboard for a while, then releasing keys for several seconds: normally, after that, the USB keyboard operation does not resume on the MS-6712.
Later, I'll test the first patch on ASUS P5QC (for a wide range of LGA775 processors up to the latest C2D/C2Q generations), which I see has an UHCI controller as well, and the second patch on the three MBs, hopefully.

**Comment by debrouxl** [comment]
When both patches are applied, I can read "kbd 0 status 1d000007", whether the keypresses are taken into account or not. Beyond a dozen keypresses or so, the key which is (sometimes) echo'ed is a blank character, space or equivalent.

**Comment by martinwhitaker** [comment]
Just to check, you've seen this with different keyboards, yes?

Can you try with this patch in place of the second one, to show what the keyboard is sending:
````
--- a/system/uhci.c
+++ b/system/uhci.c
@@ -390,6 +390,16 @@ static uint8_t get_keycode(const usb_hcd_t *hcd)
             if (status & UHCI_TD_ACTIVE) continue;
 
             hid_kbd_rpt_t *kbd_rpt = &ws->kbd_rpt[kbd_idx];
+print_usb_info("mods %02x rsvd %02x keycodes %02x %02x %02x %02x %02x %02x",
+  (uintptr_t)kbd_rpt->modifiers,
+  (uintptr_t)kbd_rpt->reserved,
+  (uintptr_t)kbd_rpt->key_code[0],
+  (uintptr_t)kbd_rpt->key_code[1],
+  (uintptr_t)kbd_rpt->key_code[2],
+  (uintptr_t)kbd_rpt->key_code[3],
+  (uintptr_t)kbd_rpt->key_code[4],
+  (uintptr_t)kbd_rpt->key_code[5]
+);
 
             uint32_t error_mask = UHCI_TD_STALLED | UHCI_TD_DB_ERR | UHCI_TD_BABBLE | UHCI_TD_NAK_RX | UHCI_TD_CRC_TO | UHCI_TD_BS_ERR;
             if (~status & error_mask) {
````

**Comment by martinwhitaker** [comment]
> A computer-side software mishap in gathering or acknowledging USB packets seems more likely than multiple keyboard models all sending bad data.

Yes, that's why I wanted to confirm you see this with more than one keyboard.

**Comment by debrouxl** [comment]
I've performed other tests on the MS-6712, this time with the MS Wired Keyboard 600. On this one, the observable behaviour is more akin to keypresses swallowed and/or stuck on the memtest86+ side ?
* ![P1100819](https://user-images.githubusercontent.com/8375/165851751-160a5a0a-c38d-4213-9861-db90816fda57.JPG) shows that 0x08 ("e") keypresses stop being taken into account at some point, despite still being seen in the packet.

* ![P1100820](https://user-images.githubusercontent.com/8375/165851799-6f4bb86f-320b-4ca6-b221-dc6b91fd0e0e.JPG) and ![P1100821](https://user-images.githubusercontent.com/8375/165851838-7853c534-c562-4d1f-b76c-7be1bb2b19f1.JPG) show a stuck 0x15 keypress, whichever other key I press. The 01 01 01 01 01 01 packets occur upon major keypress spam, then it's back to stuck mode with that 0x15 keypress. F1 / 1 is not taken into account.

**Comment by martinwhitaker** [comment]
Please test if this patch fixes it:
````
--- a/system/uhci.c
+++ b/system/uhci.c
@@ -393,6 +393,8 @@ static void poll_keyboards(const usb_hcd_t *hcd)
                 hid_kbd_rpt_t *prev_kbd_rpt = &ws->prev_kbd_rpt[kbd_idx];
                 process_usb_keyboard_report(hcd, kbd_rpt, prev_kbd_rpt);
                 *prev_kbd_rpt = *kbd_rpt;
+
+                write32(&kbd_td->token, read32(&kbd_td->token) ^ UHCI_TD_DT(1));
             }
 
             // Reenable the TD.
````

**Comment by debrouxl** [comment]
Hmm... this patch does not apply. As of 8284e987614407c4ba074757e714aade474c7beb , there are no `poll_keyboards` or `process_usb_keyboard_report` functions in that file or anywhere else in tree. There's a "Reenable the TD" comment in `get_keycode`, so I applied the line there, and it even builds, so I'll test it, but... it might well not be a good test :)

**Comment by martinwhitaker** [comment]
Oops, yes, I've not pushed my other work yet. Applying it in `get_keycode` just before the
````
             }
 
             // Reenable the TD.
````
will be a valid test.

**Comment by debrouxl** [comment]
As of af2feb78d5f54e0f37afebd4dfdb5a461ebf08ef (current spd_decoding on top of current main), this patch (number 4 in this issue) seems to solve the problem on the MS-6712 KT4AV & MS-6398E mentioned in the issue description. I have unapplied patch 2 (pair of #if 0 + trace) but left the two other tracing patches applied.
I can make slow keypresses, spam the keyboard, stop spamming the keyboard, resume keypresses... the MS Wired Keyboard 600 works as it should and communication hasn't started failing over minutes of testing, both during the initial usbdebug phase and the main memtest86+ UI. I haven't noticed slow keypresses being lost, either. Good work :)

For good measure, I'll test on the P5QC with Logitech K120 and VC820 with HP KU-0316, maybe not tonight, but I don't expect issues, given how much better the two other computers and the other keyboard behave together.

**Comment by debrouxl** [comment]
On the VC820 + KU-0316, it looks good as well. Keypresses seem to be lost during the initial usbdebug phase with an excessively long sleep (which might disturb this keyboard model ?), but no keypress was obviously lost during normal memtest86+ operation, which is what matters.
The P5QC + K120 are in another location, so tests on that pair are a task for another day.

**Comment by martinwhitaker** [comment]
Thanks. I've pushed this fix, plus the equivalent fix for the EHCI driver and (hopefully) a fix for #48 (which is what revealed the bug), so we can get wider testing.
